File: Haskel/map_fold/map_fold.cs
// Program.cs
// Select / Aggregate を使って map / fold を再現

using System;
using System.Linq;

class Program
{
    static void Main()
    {
        Console.WriteLine("=== C# map / fold (LINQ) サンプル ===");

        var numbers = new[] {1, 2, 3, 4};

        // ----------------------------------------
        // map: 各要素を2倍にする
        // ----------------------------------------
        var doubled = numbers
            .Select(x => x * 2);

        Console.WriteLine("Select(x => x * 2) = [" + string.Join(", ", doubled) + "]");

        // ----------------------------------------
        // fold: 合計を求める
        // ----------------------------------------
        var total = numbers
            .Aggregate(0, (acc, x) => acc + x);

        Console.WriteLine("Aggregate sum = " + total);

        // ----------------------------------------
        // 組み合わせ：map → fold
        // ----------------------------------------
        var result = numbers
            .Select(x => x * 2)
            .Aggregate(0, (acc, x) => acc + x);

        Console.WriteLine("sum of doubled = " + result);

        // ----------------------------------------
        // 別の例：文字列結合
        // ----------------------------------------
        var words = new[] {"Hello", " ", "World"};

        var sentence = words
            .Aggregate("", (acc, x) => acc + x);

        Console.WriteLine("文字列結合 = " + sentence);
    }
}

File: Haskel/map_fold/map_fold.hs
-- Main.hs
-- map と fold を一緒に理解するサンプル

module Main where

main :: IO ()
main = do
    putStrLn "=== Haskell map / fold サンプル ==="

    let numbers = [1,2,3,4]

    -- ----------------------------------------
    -- map: 各要素を2倍にする
    -- ----------------------------------------
    let doubled = map (*2) numbers
    putStrLn ("map (*2) [1,2,3,4] = " ++ show doubled)

    -- ----------------------------------------
    -- fold: 合計を求める
    -- ----------------------------------------
    let total = foldl (+) 0 numbers
    putStrLn ("foldl (+) 0 [1,2,3,4] = " ++ show total)

    -- ----------------------------------------
    -- 組み合わせ：map → fold
    -- ----------------------------------------
    let result = foldl (+) 0 (map (*2) numbers)
    putStrLn ("sum (map (*2) numbers) = " ++ show result)

    -- ----------------------------------------
    -- 別の例：文字列結合
    -- ----------------------------------------
    let wordsList = ["Hello", " ", "World"]
    let sentence = foldl (++) "" wordsList
    putStrLn ("文字列結合 = " ++ sentence)
